### ISSUE TITLE
builder.standardBuildOptions(.{}) 

### ISSUE THREAD

**Issue by daurnimator** [ziglang/zig#7201]
We currently have ` .standardTargetOptions()` which adds a build option to select the architecture, os and ABI. However it doesn't allow the user to e.g. enable/disable cpu features.

With the new pie support, to expose it in build.zig, we're going to want a common 'standard' option for it. e.g. `-Dpie=true`

As we continue, there's going to be more build options that projects should expose by default.

I propose that we wrap up all these commonly exposed build options in a function `.standardBuildOptions(.{})`, which can gain options over time.

**Comment by andrewrk** [comment]
I want to clarify the purpose of this function: it's for communicating to the zig build system a set of preferences or available options. For example, a set of supported targets, or the preferred release mode. It should communicate to zig build system what options are supported, what should be considered to be default, and then the build system takes that and presents the CLI options in a standard way.

Let's call it `standardOptions` since we are already in a "build" context. The preferred release mode should be deprecated and rolled into this as well.

Taking pie as an example, the person writing the build script should have to pass `.supports_pie = true` to `standardOptions` because they will have to remember to then set `.pie = true` on the build artifacts that should be affected by this option.

Or, I would also be OK with having `supports_pie` default to `true` and we could have something like this:

```zig
const opts = b.standardOptions(.{});

// ...

const exe = b.addExecutable("abc", "src/main.zig");
exe.applyOptions(opts);
```

Since `build.zig` is declarative, we can add some handy flags to the build runner for package maintainers to detect if options are supported. For example, we could add a `--list-options` flag that dumps the list of supported options to stdout in a parseable text format.

**Comment by daurnimator** [comment]
> ```zig
> const exe = b.addExecutable("abc", "src/main.zig");
> exe.applyOptions(opts);
> ```

Why not have `.addExecutable` inherit defaults from the builder context?
I might throw up a quick PR doing this for `standardReleaseOptions` now to see what you think

**Comment by andrewrk** [comment]
This was accepted, but after thinking it over some more, I'm changing status to rejected. In general, I want applications to control their options, and explicitly offer configurations to users and package maintainers via build options. Instead of an opaque `standardOptions()` function, the application should explicitly ask for a target, an optimization mode, and so on. If it wants to offer PIE, then it should explicitly offer PIE. Package maintainers should negotiate with upstream application authors on the advantages or disadvantages of PIE, and application authors can oblige or not depending on whether they expose the option.

**Comment by daurnimator** [comment]
> If it wants to offer PIE, then it should explicitly offer PIE.

As it is, we currently patch out build.zig for various zig programs we package. e.g. https://github.com/archlinux/svntogit-community/blob/packages/ncdu/trunk/ncdu-pie.patch

> Package maintainers should negotiate with upstream application authors on the advantages or disadvantages of PIE, and application authors can oblige or not depending on whether they expose the option.

We shouldn't assume application authors know about best practices. They're going to do what `zig build-exe` gives them out of the box and iterate on it.
This ends up meaning that at *some* point a packager is going to have to explain to them what PIE is and why they should care; and I would expect this happens when a package gets popular enough to be included in a common distro.

**Comment by andrewrk** [comment]
I don't think we should assume that every application should have PIE by default. ncdu certainly does not benefit from it. Of course, as the downstream, Archlinux is free to disagree with ncdu's decision, as well as zig's decision, and patch ncdu, or patch zig to enable PIE for every executable. I think a patch is the correct way to model such a disagreement, when the application developer does not endorse the decision by adding an option, and I certainly do not endorse the decision by adding it as an option to every application.

**Comment by andrewrk** [comment]
Reopening for reconsideration. I still agree with everything I said above, however, I think it's worth accepting the reality that there are a set of options that distributions do want to exert control over, and that they *will* patch if necessary. If we provide a way for distros to do it without patching, then we can also allow upstream application developers to be aware of those choices and react to them. This is more beneficial for everyone than if it is done via patching.
